_posts/2019-01-14-Infrastructure-as-code-with-terraform-and-aws-serverless.md

```diff
@@ -0,0 +1,502 @@
+
+# IAM
+<div style="text-align: center;">
+  <img src="/img/2019-01-14-Infrastructure-as-code-with-terraform-and-aws-serverless/icon-IAM.png" width="15%" height="15%">
+</div>
+
+Before specifying the Lambda Functions we have to create permissions for our functions to use.
+This makes sure that our functions have permissions to access other resources (like DynamoDB).
+Without going to deep into it, the AWS permission model works as follows:
+* Provide a resource with a role
+* Add permissions to this role
+* These allow the role to access other resources:
+    * permissions for triggering another resource (eg. Lambda Function forwards logs to CloudWatch)
+    * permissions for being triggered by another resource (eg. Lambda Function may be triggered by API Gateway)
+
+```hcl-terraform
+# ROLES
+# IAM role which dictates what other AWS services the Lambda function
+# may access.
+resource "aws_iam_role" "lambda-iam-role" {
+  name = "codingtips_lambda_role"
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "lambda.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+EOF
+}
+
+# POLICIES
+resource "aws_iam_role_policy" "dynamodb-lambda-policy"{
+  name = "dynamodb_lambda_policy"
+  role = "${aws_iam_role.lambda-iam-role.id}"
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "dynamodb:*"
+      ],
+      "Resource": "${aws_dynamodb_table.codingtips-dynamodb-table.arn}"
+    }
+  ]
+}
+EOF
+}
+```
+
+In the example above the first resource that is defined is an `aws_iam_role`.
+This is the role that we will later give to our Lambda Functions.
+
+We then create the `aws_iam_role_policy` resource which we link to the `aws_iam_role`.
+The first `aws_iam_role_policy` is giving this role permission to invoke any action on the specified DynamoDB resource.
+The second `role_policy` allows a resource with this role to send logs to CloudWatch.
+
+A couple of things to notice:
+* The `aws_iam_role` and the `aws_iam_role_policy` are connected by the `role` argument of the role_policy resource
+* In the `statement` attribute of the `aws_iam_role_policy` we grant (Effect attr.) permission to do some actions (Action attr.) on a certain Resource (Resource attr.)
+* A resource is referenced by its *arn* or *Amazon Resource Name* which uniquely identifies this resource on aws
+* There are two ways to specify an `aws_iam_role_policy`: 
+    * using the *until EOF* syntax (like I did here)
+    * using a separate Terraform `aws_iam_policy_document` element that is coupled to the `aws_iam_role_policy`
+* The dynamodb-lambda-policy allows all actions on the specified DynamoDB resource because under the action attribute it states `dynamodb:*`
```

Comment:
```suggestion
* The `dynamodb-lambda-policy` allows all actions on the specified DynamoDB resource because under the `Action` attribute it states `dynamodb:*`
```